click/core.py

```
@@ -572,6 +594,32 @@ def forward(*args, **kwargs):
 
         return self.invoke(cmd, **kwargs)
 
+    def set_parameter_source(self, name, source):
+        """Sets the `source` of a parameter. This indicates the 
+        location from which the value of the parameter was obtained.
+
+        :param name: the name of the command line parameter
+        :param source: the source of the the command line parameter,
+                       which should be a valid ParameterSource value
+        """
+        ParameterSource.validate(source)
+        self._parameter_sources[name] = source
+
+    def get_parameter_source(self, name):
+        """Get the `source` of a parameter. This indicates the 
+        location from which the value of the parameter was obtained.
+        This can be useful for determining when a user specified
+        an option on the command line that is the same as the default.
+        In that case, the source would be ParameterSource.COMMANDLINE,
+        even though the value of the parameter was equivalent to the
+        default.
+
+        :param name: the name of the command line parameter
+        :returns: the source
+        :rtype: ParameterSource
+        """
+        return self._parameter_sources.get(name)
```

`        return self._source_by_optname[name]`

(raise `KeyError` if invalid name given rather than returning `None`)